docs/typescript.md

```diff
@@ -0,0 +1,304 @@
+}
+```
+
+Each type class in TypeScript must have an associated global variable which maps unique representations of its instance types to the corresponding object of the type class implementation.
+For example, the `Eq` type class has the [global variable](https://github.com/mlabs-haskell/lambda-buffers/blob/main/runtimes/typescript/lbr-prelude/src/LambdaBuffers/Eq.ts#L11) defined in the [lbr-prelude](https://github.com/mlabs-haskell/lambda-buffers/tree/main/runtimes/typescript/lbr-prelude) library defined as follows
+
+```ts
+export const Eq: EqInstances = { } as EqInstances
+```
+
+where `EqInstances` is an interface type that is initially empty but will be extended with instances of types later.
+
+```ts
+export interface EqInstances { }
+```
+
+Finally, the following invariant is maintained in the code generator:
+
+- Every type `T` has an associated unique [symbol](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Symbol) also called `T`.
+
+So, the type `Integer` has
+
+```ts
+export type Integer = bigint
+export const Integer: unique symbol = Symbol('Integer')
+```
+
+and implementing its `Eq` instance amounts to the following code.
+
+```ts
+export interface EqInstances {
+    [Integer]: Eq<Integer>
+}
+Eq[Integer] = { eq: (l,r) => l === r
+              , neq: (l,r) => l !== r
+              }
+```
+
+For types defined in the LambdaBuffers schema, this chunk of code will be automatically generated provided there is an appropriate `derive` construct.
+
+### Type instances with constraints
+Recall in Haskell that the `Eq` instance for a tuple may be defined as follows
+
+```haskell
+instance (Eq a, Eq b) => Eq (MyPair a b) where
+    MyPair a1 a2 == MyPair b1 b2 = a1 == b1 && a2 == b2
+    MyPair a1 a2 != MyPair b1 b2 = a1 != b1 || a2 != b2
+```
+
+The corresponding TypeScript type definition and instance would be defined as follows
+
+```ts
+export type MyPair<a, b> = [a, b]
+export const MyPair: unique symbol = Symbol('MyPair')
+
+export interface EqInstances {
+    [MyPair]: <A,B>(a : Eq<A>, b : Eq<B>) => Eq<MyPair<A,B>>
+}
+Eq[MyPair] = (dictA, dictB) => { return { eq: (a,b) => dictA.eq(a[0], b[0]) && dictB.eq(a[1], b[1])
+                                        , neq: (a,b) => dictA.neq(a[0], b[0]) || dictB.neq(a[1], b[1])
+                                        } 
+                               }
+```
+
+Note that the constraints `(Eq a, Eq b) =>` become arguments `dictA` and `dictB` that are used to construct the `Eq` instance for `MyPair`.
+
+This loosely follows the original translation given in the paper [How to make ad-hoc polymorphism less ad hoc](https://dl.acm.org/doi/10.1145/75277.75283) with some minor modifications.
+
+## Limitations
+
+- Only Haskell 2010 typeclasses are supported for the TypeScript code generator. So, the following schemas will probably generate incorrect code.
+
+  ```purescript
+  derive Eq a => Eq (MyPair a a)
+  ```
```

Comment:
I might be trolling, but I can't get this example to compile with LB (it fails in the compiler phase).

## Attempts at getting it to compile
The following doesn't work:
```
module Document

-- Importing types
import Prelude (Eq)

sum MyMaybe a = MyJust a | MyNothing

derive Eq (MyMaybe (MyMaybe a))

```

```shell
$ nix develop github:mlabs-haskell/lambda-buffers#lb
$ lbf-prelude-to-haskell  Document.lbf
Success from: /nix/store/ymnlwxbl94j5ps5vs7w53xpgwhj070am-lambda-buffers-compiler-exe-lbc-0.1.0.0/bin/lbc compile --input-file .work/compiler-input.pb --output-file .work/compiler-output.pb [INFO]
/nix/store/ymnlwxbl94j5ps5vs7w53xpgwhj070am-lambda-buffers-compiler-exe-lbc-0.1.0.0/bin/lbc::0.0-0.0 When solving a constraint Prelude.Eq (MyMaybe
                                               (MyMaybe
                                               a)) no rule was found (`derive` or `instance`) to solve a goal Prelude.Eq (MyMaybe
                                                                                                                          a) in module 'Document' [ERROR]

```

Adding the instance creates overlapping instances:
```
module Document

-- Importing types
import Prelude (Eq)

sum MyMaybe a = MyJust a | MyNothing

derive Eq (MyMaybe (MyMaybe a))
derive Eq (MyMaybe a)

```
```
$ lbf-prelude-to-haskell  Document.lbf
Success from: /nix/store/ymnlwxbl94j5ps5vs7w53xpgwhj070am-lambda-buffers-compiler-exe-lbc-0.1.0.0/bin/lbc compile --input-file .work/compiler-input.pb --output-file .work/compiler-output.pb [INFO]
/nix/store/ymnlwxbl94j5ps5vs7w53xpgwhj070am-lambda-buffers-compiler-exe-lbc-0.1.0.0/bin/lbc::0.0-0.0 When solving a constraint Prelude.Eq (MyMaybe
                                               (MyMaybe
                                               a)) multiple rules were found (`derive` or `instance`) that could be used to solve a goal Prelude.Eq (MyMaybe
                                                                                                                                                     (MyMaybe
                                                                                                                                                     a)) in module 'Document'. The overlapping rules found are
 Prelude.Eq (MyMaybe (MyMaybe a)) defined in module 'Document'
Prelude.Eq (MyMaybe a) defined in module 'Document' [ERROR]
```

Finally, note that `derive Eq (Maybe a) => Eq (Maybe (Maybe a))` is not valid lbf syntax.

## Discussion
While this is an interesting example, it seems this doesn't work on all code generators because it fails the compiler phase, so I'm a bit hesitant to include this example. 

Unless I'm using this wrong? Happy to hear your thoughts!

Reply:
Yeah the compiler does say what I'd expect it to say right?

derive Eq (MyMaybe (MyMaybe Int))
derive Eq (MyMaybe (MyMaybe Bool))
instance MyMaybe Int
instance MyMaybe Bool

should probably work.

I think this is mostly a fun exercise without much use in practice, as you'd just derive the most generic type expression.

Reply:
Oooh awesome, yeah this is a great example -- I'll toss it in. This really won't work because
- the instance is not of the form `MyMaybe a` (Haskell 2010 requires the instances to be `TypeConstructor tyVar1 tyVar2` ... )
- And also, the code generator is a bit messed up when trying to compile the instance since `MyMaybe` has (in a sense) two different instance dictionaries and it doesn't know which to pick to map it to something.